cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java

```diff
@@ -296,12 +296,14 @@ private Asset getAssetForEndpoint(EndPoint endPoint) {
 
     private Map<Asset, Set<Asset>> getAssetsMapFromEndpointFieldsMap(Map<EndPointField, Set<EndPointField>>
                                                                        endPointFieldSetMap) {
-      return endPointFieldSetMap.entrySet().stream()
-        .collect(Collectors.toMap(entry -> getAssetForEndpoint(entry.getKey().getEndPoint()),
-                                  entry -> new HashSet<>(entry.getValue().stream()
-                                                           .map(endPointField ->
-                                                                  getAssetForEndpoint(endPointField.getEndPoint()))
-                                                           .collect(Collectors.toList()))));
+      Map<Asset, Set<Asset>> responseMap = new HashMap<>();
+      endPointFieldSetMap.entrySet().stream()
+        .forEach(entry -> responseMap.putIfAbsent(
+          getAssetForEndpoint(entry.getKey().getEndPoint()),
+          new HashSet<>(entry.getValue().stream()
+                          .map(endPointField -> getAssetForEndpoint(endPointField.getEndPoint()))
+                          .collect(Collectors.toList()))));
```

Comment:
Why not just use a `Collectors.toSet()`?

Reply:
Also, you can just a merging operator in your original logic to ignore values of the same key.

```
      return endPointFieldSetMap.entrySet().stream()
        .collect(Collectors.toMap(entry -> getAssetForEndpoint(entry.getKey().getEndPoint()),
                                  entry -> entry.getValue().stream()
                                                           .map(endPointField ->
                                                                  getAssetForEndpoint(endPointField.getEndPoint()))
                                                           .collect(Collectors.toSet()),
                                  (first, second) -> first));
```

Reply:
This is better as this way we will avoid creating a response map which could add to the memory in case of complex pipelines